File: graph/prim_mst_via_priority_queue.cpp
/**
 * @file prim_mst_via_priority_queue.cpp
 * @brief Prim's Minimum Spanning Tree Algorithm using Priority Queue (Adjacency List) built with using C++.
 */
//

#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <unordered_map>
#include <string>

// Define a custom comparator for priority queue
struct Compare {
    bool operator()(const std::tuple<double, std::string, std::string>& edge1,
                    const std::tuple<double, std::string, std::string>& edge2) {
        // Compare the weights of the edges and return if the weight of the first edge is greater than the second's
        return std::get<0>(edge1) > std::get<0>(edge2);
    }
};

// Function to find the Minimum Spanning Tree using Prim's Algorithm
// The graph is represented as <startingVertex, <destinationVertex, weight>>
template <typename EdgeWeightType>      // WeightType is the type of the weight of the edge (int, double, float, etc.)
void primMST(const std::unordered_map<std::string, std::vector<std::pair<std::string, EdgeWeightType>>>& graphAdjacencyList,
             const std::string& startingVertex) {

    // Priority queue to store the edges
    std::priority_queue<std::tuple<EdgeWeightType, std::string, std::string>, 
                        std::vector<std::tuple<EdgeWeightType, std::string, std::string>>, 
                        Compare> priorityQueue;

    // Track the visited vertices for MST
    std::unordered_map<std::string, bool> visitedVertices;

    // Store the weights of the MST edges
    std::vector<std::tuple<std::string, std::string, EdgeWeightType>> mstEdges;

    // Total weight of the MST
    EdgeWeightType mstTotalWeight = 0;

    // Initialize the starting vertex and its adjacencies
    visitedVertices[startingVertex] = true;
    for (const auto& neighbour : graphAdjacencyList.at(startingVertex)) {
        // (weight, startingVertex, destinationVertex)
        priorityQueue.push({neighbour.second, startingVertex, neighbour.first});
    }

    // Loop until the priority queue is empty
    while (!priorityQueue.empty()) {
        EdgeWeightType edgeWeight;
        std::string startingVertex, destinationVertex;
        std::tie(edgeWeight, startingVertex, destinationVertex) = priorityQueue.top();
        priorityQueue.pop();

        // If the destination vertex is already visited, skip the edge
        if (visitedVertices[destinationVertex])
            continue;

        // Include the edge in the MST
        visitedVertices[destinationVertex] = true;
        mstEdges.push_back(std::make_tuple(startingVertex, destinationVertex, edgeWeight));
        mstTotalWeight += edgeWeight;

        // Add the destination vertex's adjacencies to the priority queue
        for (const auto& destinationVertexNeighbour : graphAdjacencyList.at(destinationVertex)) {
            if (!visitedVertices[destinationVertexNeighbour.first]) {
                priorityQueue.push({destinationVertexNeighbour.second, destinationVertex, destinationVertexNeighbour.first});
            }
        }
    }

    // Print
    std::cout << "Minimum Spanning Tree Edges:\n";
    for (const auto& edge : mstEdges) {
        std::cout << "(" << std::get<0>(edge) << ", " << std::get<1>(edge) << ") -> Cost: " << std::get<2>(edge) << "\n";
    }
    std::cout << "Total Weight of MST: " << mstTotalWeight << "\n";
}

int main(void) {
    std::unordered_map<std::string, std::vector<std::pair<std::string, int>>> graph{
        {"A", {{"B", 2}, {"D", 6}}},
        {"B", {{"A", 2}, {"C", 3}, {"D", 8}, {"E", 5}}},
        {"C", {{"B", 3}, {"E", 7}}},
        {"D", {{"A", 6}, {"B", 8}, {"E", 9}}},
        {"E", {{"B", 5}, {"C", 7}, {"D", 9}}}
    };
    
    primMST(graph, "A");

    return 0;
}

// (A, B) -> Cost: 2
// (B, C) -> Cost: 3
// (B, E) -> Cost: 5
// (A, D) -> Cost: 6
// Total Weight of MST: 16

File: graph/prim_mst_via_priority_queue.py
# @file: prim_mst_via_priority_queue.py
# @brief: Prim's Minimum Spanning Tree Algorithm implementation in Python using priority queue to represent the graph
#         Since the grpah uses priority queue, the time complexity of the algorithm is O(E + VlogV) where V is the number of vertices in the graph
#         and E is the number of edges in the graph

from typing import List, Tuple, Dict
import heapq

def prim_mst_via_priority_queue(graph_adjacency_list) -> Tuple[List[Tuple[int, int, int]], int]:
    # Select an arbitrary vertex as the starting vertex
    start_vertex: int = next(iter(graph_adjacency_list))

    # Initialize the priority queue with the starting vertex
    priority_queue: List[Tuple[int, int]] = []

    # Keep track of the vertices that are already visited
    visited_vertices: List[bool] = [False] * len(graph_adjacency_list)

    # Keep track of the edges in the MST (The edges that connect the vertices in the MST)
    mst_edges: List[Tuple[int, int]] = []

    # Keep track of the total weight of the MST
    total_mst_weight: int = 0

    # Add the edges of the starting vertex to the priority queue
    visited_vertices[start_vertex] = True
    for neighbour_vertex, edge_weight in graph_adjacency_list[start_vertex]:
        heapq.heappush(priority_queue, (edge_weight, start_vertex, neighbour_vertex))

    # Iterate until the priority queue is empty
    while priority_queue:
        weight, source_vertex, destination_vertex = heapq.heappop(priority_queue)

        # If the destination vertex is already visited, skip the edge
        if visited_vertices[destination_vertex]:
            continue

        # Include the edge in the MST
        mst_edges.append((source_vertex, destination_vertex, weight))
        total_mst_weight += weight
        visited_vertices[destination_vertex] = True

        # Add the edges of the destination vertex to the priority queue
        for neighbour_vertex, edge_weight in graph_adjacency_list[destination_vertex]:
            if not visited_vertices[neighbour_vertex]:
                heapq.heappush(priority_queue, (edge_weight, destination_vertex, neighbour_vertex))

    return mst_edges, total_mst_weight

if __name__ == "__main__":
    # Graph is represented as adjacency list
    # graph[node] = [(neighbour, weight), ...]   (A node has a list of neighbours with their corresponding edge weights)
    graph_adjacency_list: Dict[int, List[Tuple[int, int]]] = {
        0: [(1, 2), (3, 6)],
        1: [(0, 2), (2, 3), (3, 8), (4, 5)],
        2: [(1, 3), (4, 7)],
        3: [(0, 6), (1, 8), (4, 9)],
        4: [(1, 5), (2, 7), (3, 9)]
    }

    mst_edges, total_mst_weight = prim_mst_via_priority_queue(graph_adjacency_list)
    print("Minimum Spanning Tree Edges:")
    for edge in mst_edges:
        print(f"(Vertex {edge[0]} - Vertex {edge[1]}) with weight {edge[2]}")
    print(f"Total Weight of the MST: {total_mst_weight}")

# Minimum Spanning Tree Edges:
# (Vertex 0 - Vertex 1) with weight 2
# (Vertex 1 - Vertex 2) with weight 3
# (Vertex 1 - Vertex 4) with weight 5
# (Vertex 0 - Vertex 3) with weight 6
# Total Weight of the MST: 16